### ISSUE TITLE
Generated unit names change in replays

### ISSUE THREAD

**Issue by stevecotton** [wesnoth/wesnoth#5148]
### Game and System Information
 - Local build of 1.15.4 (testing with a build directly from the tagged commit)
 - GCC 10 Running on Debian GNU/Linux bullseye/sid
 
### Describe the bug
Randomly-generated unit names change between playing the game and watching a replay.

### To Reproduce

1. Start a new game, for example A Tale of Two Brothers.
2. Recruit a unit, and remember its name.
3. Save the game.
4. Load the game, ticking the "show replay" checkbox.
5. Watch the unit being recruited, and then look at its name. The name has been generated again, instead of matching the original.
6. Repeat 4. and 5. again, see that it's got yet another name.

Note: this bug doesn't affect the loyal units that you have at the start of AToTB; all of these have their names set in the scenario.cfg file.

### Expected behavior
Names being useful identifiers for discussing replays.

### Additional context
Add any other context (compiler logs, console output, etc) about the problem here.

Game paths
==========

Data dir:        /home/USER/development/other-source/wesnoth/build_from_tag_1.15.4
User config dir: /home/USER/.config/wesnoth1.15/
User data dir:   /home/USER/.local/share/wesnoth/1.15
Saves dir:       /home/USER/.local/share/wesnoth/1.15/saves
Add-ons dir:     /home/USER/.local/share/wesnoth/1.15/data/add-ons
Cache dir:       /home/USER/.cache/wesnoth

Libraries
=========

Boost:             1.71
OpenSSL/libcrypto: 1.1.1g (runtime 1.1.1g)
Cairo:             1.16.0 (runtime 1.16.0)
Pango:             1.46.1 (runtime 1.46.1)
SDL:               2.0.12 (runtime 2.0.12)
SDL_image:         2.0.5 (runtime 2.0.5)
SDL_mixer:         2.0.4 (runtime 2.0.4)
SDL_ttf:           2.0.15 (runtime 2.0.15)

Features
========

JPEG screenshots:               yes
Lua console completion:         yes
Legacy bidirectional rendering: yes
D-Bus notifications back end:   yes

Installed add-ons
=================

After_the_Storm:        0.10.5
Ageless_Era:            4.27.0
Ageless_Resources:      4.23.0
An_Orcish_Incursion:    1.0.0
AtS_Music:              0.3.2
Black_Cross_of_Aleron:  2.0.3
Color_Modification:     1.2.1
Era_of_Magic_Resources: 1.1.1
Gui_Debug_Tools:        0.9.11
SXCollection:           0.0.0
Tactics_Puzzles:        0.3.5a

**Comment by gfgtdf** [comment]
I cannot reproduce this on current master

**Comment by gfgtdf** [comment]
Wait, actually i can but it depends on the  campaign. When i tested The hammer of Turdusan for example this problem did not occur.

**Comment by gfgtdf** [comment]
okay this seems to be not a problem for drawves also fori other units the generated name when playign the replay is not totally random but rather rotates between handful names.

**Comment by gfgtdf** [comment]
Okay i looked into this and while i don't understand the code completely i think this mutable variable https://github.com/wesnoth/wesnoth/blob/60fd504caf46bbb1648de369303b76225d805e5d/src/utils/context_free_grammar_generator.hpp#L30 looks suspicious

**Comment by CelticMinstrel** [comment]
The generator shouldn't even _run_ when viewing a replay. Even if that mutable variable is suspicious, it can't be the root issue here.

**Comment by stevecotton** [comment]
I'll vote +1 that it shouldn't run when viewing a replay. However, it seems that unit names aren't saved in ReplayWML's `[recruit]` tag, and it seems to be that way in the 1.12 savegame that I looked at too.

**Comment by gfgtdf** [comment]
> The generator shouldn't even run when viewing a replay

IIrc the was it's currently supposed to work i that it runs with the same random seed and thus returns the same name.

**Comment by CelticMinstrel** [comment]
Okay, that makes sense. So then the issue is that the context-free generator has additional state beyond just the random seed. Is it feasible to store that mutable variable in the replay? My understanding is that it's a way of reducing duplicates, though I'm not sure how effective or necessary it is.

I guess, if names are generated from the same random number pool as everything else, skipping over them would cause desync issues with stuff like damage rolls. However, that raises another possible fix. If the "recruit" action added the unit's name and name generators used their own number generator (or the unsynced one), then maybe that would work?

**Comment by gfgtdf** [comment]
> Okay, that makes sense. So then the issue is that the context-free generator has additional state beyond just the random seed. Is it feasible to store that mutable variable in the replay?

No, it would't be feasible, either make it not using an additional state, or store the whole name in the replay.

> My understanding is that it's a way of reducing duplicates

Im not sure  about that maybe it only to prevent it from having the same pattern too often within a single name, like Auuuk with tripe Us

**Comment by CelticMinstrel** [comment]
Yeah, that actually makes more sense. Maybe that's what it's for.

**Comment by stevecotton** [comment]
I had a closer look at the code, and there's a lot of oddness (see the commit that's auto-linked itself to this issue).

The handling of the `seed` doesn't make sense to me, it looks like a 640-bit PRNG seed, but it's really just caching 20 calls to the central RNG and then accessing them as a circular buffer of 20 numbers.

**Comment by gfgtdf** [comment]
> The handling of the seed doesn't make sense to me, it looks like a 640-bit PRNG seed, but it's really just caching 20 calls to the central RNG and then accessing them as a circular buffer of 20 numbers.

The main reason why it doesn'r use the main rng directly is that we want the code to always call the main rng the same number of times to prevent OOS if for example example the cleints use differnt languages which coudl result in the eng being calle da different number of time.

**Comment by gfgtdf** [comment]
I still think the 'store units names in replays' approach is interesting  and probably easy to implement, but i'm worrying a bit about whether it might increase the size of the replays.

**Comment by CelticMinstrel** [comment]
It would definitely increase the size of the replays, but dunno by how much. How many units are recruited in average game? Multiply that by, say, 10 or 20 and I guess that would approximate the size increase (in bytes) for uncompressed replays? I bet compression renders any size increase negligible however.

**Comment by gfgtdf** [comment]
> It would definitely increase the size of the replays, but dunno by how much. How many units are recruited in average game?

its not only  recruited units also units generated by wml.

> Multiply that by, say, 10 or 20 and I guess that would approximate the size increase (in bytes) for uncompressed replays

Certainly more, since there is always a  a bit overhead, 100 would be more realistic.

**Comment by CelticMinstrel** [comment]
> its not only recruited units also units generated by wml.

So you're saying a `[unit]` ActionWML generates a recruit action in the replay?

> Certainly more, since there is always a a bit overhead, 100 would be more realistic.

I think you may have misunderstood what I was saying. I meant that a single unit's name would probably add 10-20 bytes to an uncompressed replay, including the string "name=", the newline, and indentation. If your 100 here is an estimate of how many units are recruited in an average game, that means a 1-2kb increase in size - hardly significant, and once it's compressed I'm sure that increase will all but disappear.

**Comment by gfgtdf** [comment]
> So you're saying a [unit] ActionWML generates a recruit action in the replay?

No, i'm saying that a [unit] ActionWML generates a unit which obviously needs a name. There is no point in making this change only for recruited units.


> I think you may have misunderstood what I was saying. I meant that a single unit's name would probably add 10-20 bytes to an uncompressed replay, including the string "name=", the newline

No i'm not, i meant 100bytes _per unit_ in the simplest impleentation which creates a little more text than just name=<name>

**Comment by CelticMinstrel** [comment]
Why would it be 100 bytes per unit?
